crates/client/eth/src/state_update.rs

```diff
@@ -0,0 +1,261 @@
+use alloy::primitives::Address;
+use anyhow::Context;
+use dc_db::DeoxysBackend;
+use dc_metrics::block_metrics::block_metrics::BlockMetrics;
+use dp_convert::ToFelt;
+use dp_transactions::TEST_CHAIN_ID;
+use dp_utils::channel_wait_or_graceful_shutdown;
+use futures::StreamExt;
+use serde::Deserialize;
+use starknet_api::hash::StarkHash;
+use starknet_types_core::felt::Felt;
+use url::Url;
+
+use crate::client::StarknetCoreContract;
+use crate::{
+    client::EthereumClient,
+    utils::{convert_log_state_update, trim_hash},
+};
+
+#[derive(Debug, Clone, Deserialize, PartialEq)]
+pub struct L1StateUpdate {
+    pub block_number: u64,
+    pub global_root: StarkHash,
+    pub block_hash: StarkHash,
+}
+
+/// Get the last Starknet state update verified on the L1
+pub async fn get_initial_state(client: &EthereumClient) -> anyhow::Result<L1StateUpdate> {
+    let block_number = client.get_last_verified_block_number().await?;
+    let block_hash = client.get_last_verified_block_hash().await?;
+    let global_root = client.get_last_state_root().await?;
+
+    Ok(L1StateUpdate { global_root, block_number, block_hash })
+}
+
+/// Subscribes to the LogStateUpdate event from the Starknet core contract and store latest
+/// verified state
+pub async fn listen_and_update_state(
+    eth_client: EthereumClient,
+    backend: &DeoxysBackend,
+    block_metrics: BlockMetrics,
```

Comment:
I think we can use reference of `BlockMetrics` everywhere, it avoids unnecessary clones

Reply:
done